### PR TITLE
AFS clarification (closes #863) and add pre-commit troubleshooting

### LINKED ISSUE #863: clarification for joint sfs
maybe this is not the right place for the query, but would it be possible to  provide a  precise definition for  the coordinates for the  joint allele frequency spectrum (which I am interpeting as a site-frequency spectrum ?)  as it is printed out, in the documentation ? I was thinking something along the lines of M_{i,j}(n,m) denoting the  number of mutations observed in i copies in the sample of size n, and j copies in the sample of size m 
I am struggling a little to  understand the printout even from the simple example below; or if someone guides me to where this is computed  I could try to give it a go
```
import msprime as ms
import tskit

ts = ms.simulate( sample_size=6, Ne=10000, length=1e3, mutation_rate=1e-8, random_seed=24122)
##afs = ts.allele_frequency_spectrum( polarised=True,  span_normalise=False)
afs = ts.allele_frequency_spectrum( [[0,1,2], [3,4,5]],   polarised=True,  span_normalise=False)

tree=ts.first()

for site in tree.sites():
    for mutation in site.mutations:
        print("Mutation @ position {:.2f} over node {}".format(
            site.position, mutation.node))


print(tree.draw(format="unicode"))


print( afs) 
del ts
```
gives (after deleting the tree)

Mutation @ position 707.30 over node 6
Mutation @ position 708.94 over node 6
Mutation @ position 715.92 over node 9
Mutation @ position 972.97 over node 2

[[0. 0. 0. 0.]
 [1. 2. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 0.]]

### PR DESCRIPTION
**Pull request by petrelharp**
As discussed over in #863, make the AFS a bit more concrete.

In doing this I ran into some issues with `pre-commit`, yet again. I'm worried that pre-commit could be a significant deterrant to new folks, so I want to at least start documenting these things. The first problem was that my system updated to python3.8, and so the previous hook's shebang went looking for python3.7 but couldn't find it. Reinstalling the pre-commit hook fixed it. But, then I've got some formatting errors (black?) on files I didn't change -- see my additions to `docs/development.rst`. My advice here is just do `--no-verify`... but, what's the right way to fix this?

### CODE REVIEW

**Review by jeromekelleher — COMMENTED** [review]
Running --no-verify when you hit problems on pre-commit is good advice all right. I'm not sure the first one should be something people hit, and I'd rather solve that at source than document it as something to expect pre-commit to die on.

I agree it's a bit of an uphill one for new contributors, but them so also would be asking them to run black, etc etc, on their code as part of the PR process. There's a tradeoff here between making things easier for contributors and making things easier for maintainers.

**Comment by jeromekelleher** [comment]
The AFS things look good. I'm not so sure about the pre-commit stuff though.

**Comment by petrelharp** [comment]
> I'm not sure the first one should be something people hit, and I'd rather solve that at source than document it as something to expect pre-commit to die on.

Sure, but I've run into this problem a bunch of times by now (I think because my black wasn't up to date or something?), so we should document what to do.

**Comment by petrelharp** [comment]
I wonder if we can add a custom message to `pre-commit`? Like "It doesn't look like your code passed the pre-commit check: you should fix the errors above or append --no-verify to your git commit command."

File: docs/development.rst
Comment (benjeffery) [reply docs/development.rst:930]: Something odd is happening here. This is _defiantly_ not due to a pre-commit update. Pre-commit runs in an isolated environment with fixed versions. When I update the versions I also commit any changes as in #856. But the proof is that the error message here `E203` is specifically disabled in our  `flake8` config as it is not compatible with `black`:
`ignore = E203, E501, W503, B305`.
Somehow the `.flake8` file was not being used or read correctly.

Comment (petrelharp) [reply docs/development.rst:930]: omg, here's the problem. My previous pre-commit hook had been copied by `pre-commit install` to `.git/hooks/pre-commit.legacy`; I assumed that meant it was not being run but in fact it *was* still being run, and it was doing flake8 but without the config. I'm still not sure *why* it was being run (is this a git thing, or does pre-commit look for the legacy hook file?). Updating the docs now.

**Comment by petrelharp** [comment]
Ok - updated the pre-commit-related docs. Maybe no-one else will run into the problems I did, so this is too verbose?

File: docs/development.rst
Comment (jeromekelleher) [reply docs/development.rst:943]: Ah, that makes more sense.

**Review by jeromekelleher — APPROVED** [review]
LGTM

**Comment by jeromekelleher** [comment]
I'll let you add the merge-label here @benjeffery, in case there's changes you'd like to see.

I feel like the directions might be a little verbose and specific, but then people will probably hit this *type* of thing out of pre-commit, so after reading this they might try some of the commands which would probably sort them out.

File: docs/development.rst
Comment (benjeffery) [inline docs/development.rst:1002]: MIght be good to put the `-f` option here as that stops pre-commit keeping the old hooks.

**Review by benjeffery — APPROVED** [review]
Just one pointer on pre-commit.

File: docs/development.rst
Comment (petrelharp) [reply docs/development.rst:943]: where's the "facepalm" reaction?

File: docs/development.rst
Comment (petrelharp) [reply docs/development.rst]: done.

**Pull request merged**
